### DISCUSSION TITLE
Translation context for Labels, Buttons, Tooltips, etc

### DISCUSSION THREAD

**Discussion by ghost** [godotengine/godot-proposals#8061] [General Discussions]
As of now, Godot can generate .pot from scripts, resources, labels, buttons, etc. In scripts, Godot will collect translations from tr() and tr_n() functions, where user can provide optional 2rd (for tr) and 4th (for tr_n) string argument, which would end up as translation context hint in .pot file for string that need to be translated, like this:
```gdscript
tr("Test Label", "Some context info")
```
Will become this in .pot file:
```po
msgctxt "Some context info"
msgid "Test Label"
msgstr ""
```

But it seems, there no such option for Labels, Buttons, Tooltips, etc, that user added via Text input. At least, after some documentation reading and going through Godot ui, I can't find this option there.

I think, it could be good idea to provide additional input field for Labels, Buttons, Tooltips, etc, that would be added as translation context to that text or tooltip:
![зображення](https://github.com/godotengine/godot-proposals/assets/54743395/cfba492b-65ba-445f-8e88-01dbaf777b48)

**Comment by Calinou** [comment]
This should probably be a property on Control, similar to the Auto Translate property.

**Reply by shitake2333** [reply]
`auto_translate ` has been moved to `node`, and I think the `Localization` property group should also be moved to the `node`.

**Comment by shitake2333** [comment]
Before supporting automatic collection of localization resource from the UI system, you must first mark which properties on the UI node need to be collected.

In addition, the contents of the `Localization` property group should only be displayed on node that have multilingual resources.

**Comment by timothyqiu** [comment]
The problem I see is that translation context and translation comment should be text specific, not control specific. But the definition of "this control's text" varies.

Some control might contain multiple text fields so a single translation context won't fit all needs. Not to mention controls like `ItemList`, `Tree`, and `OptionButton` that contain multiple items, and each item could need different contexts.

If taken tooltips into account, every control's translation context also needs to be doubled.

So currently it's just easier to use `tr()` in `NOTIFICATION_TRANSLATION_CHANGED` when context is needed.

I think injecting extra elements like translation context and comment into the auto translation system needs a more in-depth design. Or we can call auto translation system simple-usecase-only, only adding `translation_context` properties to simple controls like `Label` and `Button`.

**Reply by shitake2333** [reply]
I am doing related work.

I added the `localization_resources` attribute to `node` to record the resources to be collected for internationalization on a node component.

`localization_resources` is a list of `I18nResource` :
```
struct I18nResource {
    I18nResourceType type;
    String property_name;
};
```

In this way, we can record the information of multilingual resources that need to be collected according to the actual situation of each node.

```
// TextureButton:
#if TOOLS_ENABLED
	set_has_i18n_resource(true);
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_normal");
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_pressed");
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_hover");
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_disabled");
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_focused");
	add_i18n_resource(I18N_RESOURCE_IMAGE, "texture_click_mask");
#endif
```

For more information, you can check out my new commits on the [branch](https://github.com/molingyu/godot/tree/new_localization_system)

**Reply by jkirsteins** [reply]
> The problem I see is that translation context and translation comment should be text specific, not control specific. But the definition of "this control's text" varies.
> 
> Some control might contain multiple text fields so a single translation context won't fit all needs. Not to mention controls like `ItemList`, `Tree`, and `OptionButton` that contain multiple items, and each item could need different contexts.
> 

I'm not sure I agree with this. A single ItemList should not have multiple contexts, because the translation context is the ItemList itself. Knowing where the values are displayed is enough context to translate them properly.

The only reason I can think of to having multiple contexts "per item" in some controls, would be to save translation effort. But in practice translation tools allow re-using similar translations across different contexts easily.

In my opinion, the original proposal offers good value for effort. It seems simpler to implement, and the solution is widely applicable.

**Comment by jkirsteins** [comment]
I put together a proof-of-concept of how this could be implemented: https://github.com/jkirsteins/godot/pull/4

**A concrete example** 

Before the change I use the string _General_ in two places. A label in a settings screen (as in _General settings_) and on a card (as in "this is a card of an army General").

Before the change the generated `*.pot` file (the comment shows that it's used in two files):

```
#: card_template.tscn scenes/main_menu/settings_window.tscn
#: cards/card_data.tres
msgid "General"
msgstr "??? conflicting translation ???"
```

After the change you can specify the translation context on the label right in the Editor:

![Screenshot 2025-02-21 at 00 14 41](https://github.com/user-attachments/assets/b7dc61d8-5418-45e8-bc9b-b3f6f47cd5f7)

And the generated `*.pot/*.po` files now have two items for `General`, which can be translated separately:

```
#: card_template.tscn cards/card_data.tres
msgid "General"
msgstr "Le général"

#: scenes/main_menu/settings_window.tscn
msgctxt "settings_window"
msgid "General"
msgstr "Paramètres généraux"
```

**Comment by davcri** [comment]
As of 2026, is there any way to add context to Labels and other simpler Control nodes in Godot 4.6+?

Having a way to specify translation context would be really nice.

**Reply by Calinou** [reply]
This was implemented in 4.7:

- https://github.com/godotengine/godot/pull/115340
